drivers/eeprom/eeprom_shell.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief EEPROM shell commands.
+ */
+
+#include <shell/shell.h>
+#include <drivers/eeprom.h>
+#include <stdlib.h>
+
+struct args_index {
+	u8_t device;
+	u8_t offset;
+	u8_t length;
+	u8_t data;
+	u8_t pattern;
+};
+
+struct args_number {
+	u8_t read;
+	u8_t write;
+	u8_t size;
+	u8_t fill;
+};
+
+static const struct args_index args_indx = {
+	.device = 1,
+	.offset = 2,
+	.length = 3,
+	.data = 3,
+	.pattern = 4,
+};
+
+static const struct args_number args_no = {
+	.read = 4,
+	.write = 4, /* at least 4 arguments */
+	.size = 2,
+	.fill = 5,
+};
+
+static int cmd_read(const struct shell *shell, size_t argc, char **argv)
+{
+	u8_t buf[CONFIG_EEPROM_SHELL_BUFFER_SIZE];
+	struct device *eeprom;
+	off_t offset;
+	size_t len;
+	int err;
+
+	if (argc != args_no.read) {
+		shell_help(shell);
+		return SHELL_CMD_HELP_PRINTED;
+	}
+
+	offset = strtoul(argv[args_indx.offset], NULL, 0);
+	len = strtoul(argv[args_indx.length], NULL, 0);
+
+	if (len > sizeof(buf)) {
+		shell_error(shell, "Read buffer size (%d bytes) exceeded",
+			    sizeof(buf));
+		return -EINVAL;
+	}
+
+	eeprom = device_get_binding(argv[args_indx.device]);
+	if (!eeprom) {
+		shell_error(shell, "EEPROM device not found");
+		return -EINVAL;
+	}
+
+	shell_print(shell, "Reading %d bytes from EEPROM, offset %d...", len,
+		    offset);
+
+	err = eeprom_read(eeprom, offset, buf, len);
+	if (err) {
+		shell_error(shell, "EEPROM read failed (err %d)", err);
+		return err;
+	}
+
+	shell_hexdump(shell, buf, len);
+
+	return 0;
+}
+
+static int cmd_write(const struct shell *shell, size_t argc, char **argv)
+{
+	u8_t wr_buf[CONFIG_EEPROM_SHELL_BUFFER_SIZE];
+	u8_t rd_buf[CONFIG_EEPROM_SHELL_BUFFER_SIZE];
+	struct device *eeprom;
+	unsigned long byte;
+	off_t offset;
+	size_t len;
+	int err;
+	int i;
+
+	if (argc < args_no.write) {
```

Comment:
You can use `SHELL_CMD_ARG` in lines 238-242 and pass the number of mandatory and optional parameters instead of this check here. Same for the read command.

Reply:
@alexanderwachter Thanks! I have updated the code to use `SHELL_CMD_ARG()`.